tiledb/sm/array_schema/dimension.h

```diff
@@ -141,6 +141,172 @@ class Dimension {
   /** Returns the dimension name. */
   const std::string& name() const;
 
+  /**
+   *  Returns the tile index for integer values.
+   *
+   * @param v The value.
+   * @param domain_low The minimum value for the domain.
+   * @param tile_extent The tile extent.
+   * @return The index of the tile.
+   */
+  template <
+      class T,
+      typename std::enable_if<std::is_integral<T>::value>::type* = nullptr>
+  static uint64_t tile_idx(
+      const T& v, const T& domain_low, const T& tile_extent) {
+    typedef typename std::make_unsigned<T>::type unsigned_t;
+    return ((unsigned_t)v - (unsigned_t)domain_low) / (unsigned_t)tile_extent;
+  }
+
+  /**
+   *  Returns the tile index for floating point values.
+   *
+   * @param v The value.
+   * @param domain_low The minimum value for the domain.
+   * @param tile_extent The tile extent.
+   * @return The index of the tile.
+   */
+  template <
+      class T,
+      typename std::enable_if<!std::is_integral<T>::value>::type* = nullptr>
+  static uint64_t tile_idx(
+      const T& v, const T& domain_low, const T& tile_extent) {
+    return (v - domain_low) / tile_extent;
+  }
+
+  /**
+   *  Rounds the value down to the tile boundary for integer values.
+   *
+   * @param v The value.
+   * @param domain_low The minimum value for the domain.
+   * @param tile_extent The tile extent.
+   * @return The value rounded down to the tile boundary.
+   */
+  template <
+      class T,
+      typename std::enable_if<std::is_integral<T>::value>::type* = nullptr>
+  static T round_to_tile(
+      const T& v, const T& domain_low, const T& tile_extent) {
+    typedef typename std::make_unsigned<T>::type unsigned_t;
+    return ((unsigned_t)v - (unsigned_t)domain_low) / (unsigned_t)tile_extent *
+               (unsigned_t)tile_extent +
+           (unsigned_t)domain_low;
+  }
+
+  /**
+   *  Rounds the value down to the tile boundary for floaring point values.
+   *
+   * @param v The value.
+   * @param domain_low The minimum value for the domain.
+   * @param tile_extent The tile extent.
+   * @return The value rounded down to the tile boundary.
+   */
+  template <
+      class T,
+      typename std::enable_if<!std::is_integral<T>::value>::type* = nullptr>
+  static T round_to_tile(
+      const T& v, const T& domain_low, const T& tile_extent) {
+    return (v - domain_low) / tile_extent * tile_extent + domain_low;
```

Comment:
I think this should be:
`return floor((v - domain_low) / tile_extent) * tile_extent + domain_low;`